Packet++/src/CiscoHdlcLayer.cpp

```
@@ -0,0 +1,130 @@
+	}
+
+	void CiscoHdlcLayer::computeCalculateFields()
+	{
+		if (m_NextLayer != nullptr)
+		{
+			switch (m_NextLayer->getProtocol())
+			{
+			case IPv4:
+			{
+				setNextProtocol(CISCO_HDLC_TYPE_IP);
+				break;
+			}
+			case IPv6:
+			{
+				setNextProtocol(CISCO_HDLC_TYPE_IPV6);
+				break;
+			}
+			}
+		}
+	}
+
+	void CiscoHdlcLayer::parseNextLayer()
+	{
+		auto payload = m_Data + sizeof(cisco_hdlc_header);
+		auto payloadLen = m_DataLen - sizeof(cisco_hdlc_header);
+
+		auto nextProtocol = be16toh(getCiscoHdlcHeader()->protocol);
+
+		switch (nextProtocol)
+		{
+		case CISCO_HDLC_TYPE_IP:
+		{
+			m_NextLayer = IPv4Layer::isDataValid(payload, payloadLen)
+			                  ? static_cast<Layer*>(new IPv4Layer(payload, payloadLen, this, m_Packet))
+			                  : static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+			break;
+		}
+		case CISCO_HDLC_TYPE_IPV6:
+		{
+			m_NextLayer = IPv6Layer::isDataValid(payload, payloadLen)
+			                  ? static_cast<Layer*>(new IPv6Layer(payload, payloadLen, this, m_Packet))
+			                  : static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+			break;
+		}
+		default:
+		{
+			m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
+			break;
+		}
+		}
+	}
+
+	std::string CiscoHdlcLayer::toString() const
+	{
+		return "Cisco HDLC Layer";
+	}
+
+	CiscoHdlcLayer::Address CiscoHdlcLayer::getAddress() const
+	{
+		switch (static_cast<Address>(getAddressValue()))
+		{
+		case Address::Unicast:
+		case Address::Multicast:
+		{
+			return static_cast<Address>(getAddressValue());
+		}
+		default:
+		{
+			return Address::Unknown;
+		}
+		}
+	}
+
+	uint8_t CiscoHdlcLayer::getAddressValue() const
+	{
+		return getCiscoHdlcHeader()->address;
+	}
+
+	void CiscoHdlcLayer::setAddress(Address address)
+	{
+		address = (address == Address::Unknown ? Address::Unicast : address);
+		setAddressValue(static_cast<uint8_t>(address));
+	}
```

Why are we silently converting Unknown to Unicast? Wouldn't it be better to raise `std::invalid_argument`?

---

Fixed in this commit: https://github.com/seladb/PcapPlusPlus/pull/1747/commits/e7ed1bda266eda17ae80bcda0aa429f190bdd182